### PR TITLE
[Metal] Change the function argument pointer type to use `&` instead of `*`

### PR DESCRIPTION
**Pull request by expenses**
This ~~, alongside https://github.com/gfx-rs/naga/pull/692, fully~~ fixes https://github.com/gfx-rs/naga/issues/685. ~~It feels a bit like a hacky solution though and I'm not sure if it's going to work in all cases.~~

### CODE REVIEW

**Comment by expenses** [comment]
This seems to increase the stack size of `put_block` to `13632` up from the last recorded size of `12736`.

File: src/back/msl/writer.rs [hunk 1]
```diff
@@ -1284,11 +1284,25 @@ impl<W: Write> Writer<W> {
...
+                    for (i, (&handle, argument)) in
+                        arguments.iter().zip(function_arguments).enumerate()
+                    {
                         if i != 0 {
                             write!(self.out, ", ")?;
                         }
+
+                        let ty = &context.expression.module.types[argument.ty];
+
+                        // If the function expects a pointer as an argument, take a reference.
```
Comment (kvark) [inline src/back/msl/writer.rs:1301]: Is there a reason for this to be handled specifically for function parameters? I'd expect it to be handled more generally by `put_expression` instead.

**Review by kvark — CHANGES_REQUESTED** [review]

File: src/back/msl/writer.rs (hunk 1 above)
Comment (expenses) [reply src/back/msl/writer.rs:1301]: That probably would be better, but I'm not sure how that should be done. In the trace I just have

```
[TRACE naga::back::msl::writer] statement[1] Call { function: [1], arguments: [[15], [16], [17]], result: Some([18]) }
[TRACE naga::back::msl::writer] expression [15] = LocalVariable([1])
[TRACE naga::back::msl::writer] expression [16] = LocalVariable([2])
[TRACE naga::back::msl::writer] expression [17] = LocalVariable([3])
```
and I can't see how a `LocalVariable` can know that it's meant to be referenced with `&` with the information it has available.

Comment (kvark) [reply src/back/msl/writer.rs:1301]: Perhaps the trace log is misleading? Our validation strictly requires the types of arguments to match, see https://github.com/gfx-rs/naga/blob/98252cf5d29790f85cc4397e0cf7a5b6cfaac614/src/valid/function.rs#L202
So the arguments passed into a call (of a function that expects pointers) are pointers, given that the validation succeeded.

**Comment by expenses** [comment]
The fix for this was actually pretty easy; MSL casts to references but not pointers so we just have to make the function take those instead.

**Comment by kvark** [comment]
Why are we coercing pointers to references, exactly?
This may be the way to go, I just want to understand what are the blockers for treating pointers as real pointers.

**Comment by expenses** [comment]
> Why are we coercing pointers to references, exactly?
> This may be the way to go, I just want to understand what are the blockers for treating pointers as real pointers.

As far as I can tell, references are the way to go in C++, and pointers should only be used when necessary. Using references for metal makes a ton of stuff easier, such as not requiring `&` when passing arguments and not needing the changes from https://github.com/gfx-rs/naga/pull/697 etc.

The one downside is that it makes the semantics a little different from the IR, as you're never having to deref using `*` in `Expression::Load` etc.

**Comment by kvark** [comment]
I suppose it's fine, given that we already generate a ton of references in the boilerplate code.

**Review by kvark — APPROVED** [review]

**Comment by kvark** [comment]
I ran the Dota shader conversion with this and didn't see any issues.

**Pull request merged**